src/tdlib/chat_list_item.rs

```diff
@@ -0,0 +1,117 @@
+use crate::tdlib::Chat;
+
+mod imp {
+    use super::*;
+    use glib::WeakRef;
+    use once_cell::sync::Lazy;
+    use std::cell::Cell;
+
+    #[derive(Debug, Default)]
+    pub(crate) struct ChatListItem {
+        pub(super) chat: WeakRef<Chat>,
+        pub(super) order: Cell<i64>,
+        pub(super) is_pinned: Cell<bool>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for ChatListItem {
+        const NAME: &'static str = "ChatListItem";
+        type Type = super::ChatListItem;
+    }
+
+    impl ObjectImpl for ChatListItem {
+        fn properties() -> &'static [glib::ParamSpec] {
+            static PROPERTIES: Lazy<Vec<glib::ParamSpec>> = Lazy::new(|| {
+                vec![
+                    glib::ParamSpecObject::builder::<Chat>("chat")
+                        .read_only()
+                        .build(),
+                    glib::ParamSpecInt64::builder("order").read_only().build(),
+                    glib::ParamSpecBoolean::builder("is-pinned")
+                        .read_only()
+                        .build(),
+                ]
+            });
+            PROPERTIES.as_ref()
+        }
+
+        fn property(&self, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+            let obj = self.obj();
+
+            match pspec.name() {
+                "chat" => obj.chat().to_value(),
+                "order" => obj.order().to_value(),
+                "is-pinned" => obj.is_pinned().to_value(),
+                _ => unimplemented!(),
+            }
+        }
+    }
+}
+
+glib::wrapper! {
+    pub(crate) struct ChatListItem(ObjectSubclass<imp::ChatListItem>);
+}
+
+impl ChatListItem {
+    pub(crate) fn new(chat: &Chat, position: &TdChatPosition) -> ChatListItem {
+        let obj: ChatListItem = glib::Object::builder().build();
+        let imp = obj.imp();
+
+        imp.chat.set(Some(chat));
+        imp.order.set(position.order);
+        imp.is_pinned.set(position.is_pinned);
+
+        obj
+    }
+
+    pub(crate) fn update(&self, position: &TdChatPosition) {
+        self.set_order(position.order);
+        self.set_is_pinned(position.is_pinned);
+    }
+
+    pub(crate) fn chat(&self) -> Chat {
+        self.imp().chat.upgrade().unwrap()
+    }
+
+    pub(crate) fn order(&self) -> i64 {
+        self.imp().order.get()
+    }
+
+    fn set_order(&self, order: i64) {
+        if self.order() == order {
+            return;
+        }
+        self.imp().order.set(order);
+        self.notify("order");
+    }
+
+    pub(crate) fn is_pinned(&self) -> bool {
+        self.imp().is_pinned.get()
+    }
+
+    fn set_is_pinned(&self, is_pinned: bool) {
+        if self.is_pinned() == is_pinned {
+            return;
+        }
+        self.imp().is_pinned.set(is_pinned);
+        self.notify("is-pinned");
+    }
+
+    pub(crate) async fn toggle_is_pinned(&self) -> Result<(), TdError> {
+        let chat = self.chat();
+        functions::toggle_chat_is_pinned(
+            TdChatList::Main,
```

Comment:
Oops, this needs to be changed. The `ChatListItem` should store the `ChatList` type and use that instead.